src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,346 @@
+        {
+            var heartbeat = new MetricTelemetry(heartbeatSyntheticMetricName, 0.0);
+            string updatedKeys = string.Empty;
+            string comma = string.Empty;
+
+            this.AddPropertiesToHeartbeat(heartbeat, this.sdkPayloadItems);
+            this.AddPropertiesToHeartbeat(heartbeat, this.extendPayloadItems);
+
+            heartbeat.Sequence = string.Format(CultureInfo.CurrentCulture, "{0}", this.heartbeatsSent++);
+
+            return heartbeat;
+        }
+
+        protected void SetDefaultPayloadItems()
+        {
+            HeartbeatDefaultPayload defaultPayload = new HeartbeatDefaultPayload(this.disabledDefaultFields);
+            IDictionary<string, HeartbeatPropertyPayload> defaultProps = defaultPayload.GetPayloadProperties();
+
+            this.sdkPayloadItems.Clear();
+            foreach (var kvpProp in defaultProps)
+            {
+                this.AddHealthPropertyInternal(this.sdkPayloadItems, kvpProp.Key, kvpProp.Value.PayloadValue, kvpProp.Value.IsHealthy);
+            }                
+        }
+
+        private void AddPropertiesToHeartbeat(MetricTelemetry hbeat, ConcurrentDictionary<string, HeartbeatPropertyPayload> props)
+        {
+            string updatedKeys = string.Empty;
+            string comma = hbeat.Properties.Count <= 0 ? string.Empty : ",";
+
+            lock (props)
+            {
+                foreach (var payloadItem in props)
+                {
+                    try
+                    {
+                        hbeat.Properties.Add(payloadItem.Key, payloadItem.Value.PayloadValue);
+                        hbeat.Sum += payloadItem.Value.IsHealthy ? 0 : 1;
+                        if (payloadItem.Value.IsUpdated)
+                        {
+                            string.Concat(updatedKeys, comma, payloadItem.Key);
+                            comma = ",";
+                        }
+                    }
+                    catch (Exception)
+                    {
+                        // skip sending this payload item out, no need to interrupt other payloads. Log it to core.
+                        CoreEventSource.Log.LogError(string.Format(CultureInfo.CurrentCulture, "Failed to send payload for item {0}.", payloadItem.Key));
+                    }
+                }
+            }
+
+            // update the special 'updated keys' property with the names of keys that have been updated.
+            if (!string.IsNullOrEmpty(updatedKeys))
+            {
+                hbeat.Properties[HeartbeatDefaultPayload.UpdatedFieldsPropertyKey] = updatedKeys;
+            }
+        }
+
+        private bool SetHealthPropertyInternal(ConcurrentDictionary<string, HeartbeatPropertyPayload> properties, string name, string payloadValue, bool? isHealthy)
+        {
+            try
+            {
+                properties.AddOrUpdate(name, (key) => 
+                {
+                    throw new Exception("Not allowed to set this!");
```

Comment:
How often do you expect this exception? Could it be often due to API misuse, so we can prevent this by documenting possible pitfalls in API? I worry that the extra exceptions on happy path may skew exceptions statistics for the app with AI's own exceptions.

Reply:
This will happen if a user attempts to set a heartbeat property without adding it first. I expect it to happen generally when the user is first starting out extending properties (and skips the AddHealthProperty method).

The Add-, Set-, and RemoveHealthProperty methods are fairly self-evident **_to me_** in how they would be used, and they **_are_** documented as such already (and everybody reads docs right?).

I am very open to your thoughts here though - simplify and just go with Set/Remove? More work has to be done to avoid allowing the user to collide property names but it's not that difficult to do...

**I have** set this exception to read something more meaningful though - 'Not allowed to set this without adding the health property first'.